### PR TITLE
Support for tilting gravity

### LINKED ISSUE #1151: Abstraction for domains tilted with respect to `gravitational_acceleration`
We currently assume that the vertical direction (z) is aligned with gravity, eg:

https://github.com/CliMA/Oceananigans.jl/blob/d008f382eceebfd6c4a5c5ea78cd41dc3054513d/src/TimeSteppers/update_hydrostatic_pressure.jl#L14

This is reflected by the API for `SeawaterBuoyancy`, which assumes that `gravitational_acceleration` is a scalar:

https://github.com/CliMA/Oceananigans.jl/blob/d008f382eceebfd6c4a5c5ea78cd41dc3054513d/src/Buoyancy/seawater_buoyancy.jl#L37-L42

It'd be nice to support vector `gravitational_acceleration`s that specify the gravitational acceleration in `x, y, z` independently. For this we can dispatch on whether gravitational acceleration is a `Number` to assume vertical gravity, and more generally suppor the case that `length(gravitational_acceleration) = 3`.

We can mirror this support in the API for `BuoyancyTracer`; however for that case we may want to normalize the `gravitational_acceleration` vector so that the definition of buoyancy remains the same.

This will require adding buoyancy forces to the horizontal momentum equations.

We may also want to have a discussion about using dispatch to distinguish between hydrostatic versus non-hydrostatic physics (which we do for every term _except_ pressure). Calculating the hydrostatic pressure even when the model is non-hydrostatic is both a more convoluted algorithm as well as being more memory / computationally expensive. This change could be easily dealt with in the same PR that adds buoyancy terms to the horizontal momentum equation by also adding a buoyancy term to the vertical momentum equation, setting `pHY = nothing`, and providing both 

```julia
update_hydrostatic_pressure!(::Nothing, args...) = nothing
```

and operators like

```julia
∂xᶠᵃᵃ(i, j, k, grid::AbstractGrid{FT}, ::Nothing) where FT = zero(FT) 
```

Thus switching to having a single pressure field only requires a handful of new lines, and we can keep the kernel for calculating hydrostatic pressure if/when we implement a hydrostatic-only model.

### PR DESCRIPTION
**Pull request by ali-ramadhan**
This PR adds experimental support for rotating/tilting the gravity vector.

I'm not even sure if it works properly but I'm opening a draft PR as it might be ready to be experimented with? I don't think we can have a non-zero gravity component along a periodic direction as the fluid just free falls along that dimension and the model blows up.

But here's a thermal bubble rising with gravity tilted at 45 degrees

tilted_gravity_plume

This is an experimental feature since it needs more rigorous testing and because full support for tilted gravity may require more work, e.g.
1. Should AMD use `z_dot_g_b` now?
2. Should `∂x_b` return a vector or should it be split up into `x_dot_∂x_b`, etc.? The Leith closure uses `∂x_b`.
3. We may want to clean up and refactor the implementation, taking some suggestions from #1151.

cc @tomchor 

Resolves #1151

### CODE REVIEW

**Comment by glwagner** [comment]
> I don't think we can have a non-zero gravity component along a periodic direction as the fluid just free falls along that dimension and the model blows up.

I'm confused how this is possible. Was the buoyancy initial condition periodic?

**Comment by glwagner** [comment]
> Should AMD use z_dot_g_b now?

The implementation of a "buoyancy-aware" AMD is not correct...

But yes, in principle, it would have to know the direction of gravity. We'd have to add terms that depend on the horizontal directions for it to be correct I think, which probably warrants re-deriving it for generalized gravity from scratch.

**Comment by tomchor** [comment]
@ali-ramadhan Thanks a lot for this! CCing @wenegrat here so that he's also aware.

> I'm confused how this is possible. Why would the model blow up? Was the buoyancy initial condition periodic?

@glwagner I think what he means is that there's nothing to stop the buoyant parcels if the direction is periodic (it becomes an infinite direction). So the parcels (that are buoyantly non-neutral) keep accelerating endlessly with nothing to stop them until the velocity is large enough that the model blows up.


> I don't think we can have a non-zero gravity component along a periodic direction as the fluid just free falls along that dimension and the model blows up.

@ali-ramadhan I think you can, but you need to have something else in that direction to balance it out. Like a background buoyancy gradient or drag.

**Comment by glwagner** [comment]
> Should ∂x_b return a vector or should it be split up into x_dot_∂x_b, etc.? The Leith closure uses ∂x_b.

Hmm, that's annoying. I guess you could write this `∂x_z_dot_g_b` to be perfectly correct and verbose. Or maybe we need a shorthand for the vertical component of buoyant acceleration, like `∂x_bᶻ`. I'm 99% sure we would never want to use Leith in a tilted domain.

**Comment by johncmarshall54** [comment]
Ali, good progress. Does this keep a resting fluid stratified in the
direction of gravity at rest? John

**Comment by glwagner** [comment]
> @glwagner I think what he means is that there's nothing to stop the buoyant parcels if the direction is periodic (it becomes an infinite direction). So the parcels (that are buoyantly non-neutral) keep accelerating endlessly with nothing to stop them until the velocity is large enough that the model blows up.

How is this possible with a periodic solution?

File: src/Buoyancy/gravitational_acceleration.jl
Comment (glwagner) [inline src/Buoyancy/gravitational_acceleration.jl:7]: Nice solution!

File: src/Buoyancy/linear_equation_of_state.jl
Comment (glwagner) [inline src/Buoyancy/linear_equation_of_state.jl:43]: Missing whitespace in `{FT,G}`

File: src/Buoyancy/gravitational_acceleration.jl
Comment (glwagner) [reply src/Buoyancy/gravitational_acceleration.jl:7]: Although I guess this is circumvented if we just convert `g` to a tuple in the constructor, right? This would also simplify the struct since we wouldn't need the parameters `FT` and `G` separately.

Comment (glwagner) [reply src/Buoyancy/gravitational_acceleration.jl:7]: May need `@inbounds`.

**Comment by tomchor** [comment]
> How is this possible with a periodic solution?

Maybe I'm missing something, but I can't imagine how that would _not_ happen. Take the example animation that @ali-ramadhan posted in the first message. If you image the same IC but with a domain periodic in x and z the blob would accelerate towards the upper right and come back in the lower left endlessly, but constantly gaining speed.

**Comment by glwagner** [comment]
> Maybe I'm missing something, but I can't imagine how that would not happen. Take the example animation that @ali-ramadhan posted in the first message. If you image the same IC but with a domain periodic in x and z the blob would accelerate towards the upper right and come back in the lower left endlessly, but constantly gaining speed.

I think the problem is that we cannot impose a `b` that has a non-zero domain average. This is consistent with its definition (eg it is proportional to the density _perturbation_; we subtract out a reference density in Boussinesq).

Consider the x-momentum equation:

```
du/dt + ... = b sin(theta)
```

if `b` has no net average, then the average of this equation is just

```
d <u> / dt = 0
```

right?

**Comment by glwagner** [comment]
The misleading part is this comment:

> I don't think we can have a non-zero gravity component along a periodic direction as the fluid just free falls along that dimension and the model blows up.

There's a lot of research that uses tilted domains with periodic directions (I think JO Wenegrat, LN Thomas, J Gula, JC McWilliams 2018 is one example?) --- so either that body of research is wrong, or this comment is wrong. I think it is the latter.

**Comment by glwagner** [comment]
I think maybe the main point is that periodic directions with at tilted domain are almost certainly okay. The buoyancy field represents a _density perturbation_ from its mean value --- in other words, we certainly _are_ subtracting out a mean contribution to the pressure field (`rho_0 g z` in the case that gravity aligns with gravity). We want to continue to do this when the domain is tilted with respect to gravity. With appropriate initial conditions, the fluid will not "accelerate indefinitely".

It is nevertheless possible to impose initial conditions that blow up in tilted domains (like in any other problem...) In particular, if we impose a buoyancy field with a non-zero mean, we are effectively adding a net force to the momentum equation in the periodic direction. I think this is usually unintentional --- its akin to adding a background pressure gradient. Probably, the physical problem that one wanted to solve in this case had zero mean buoyancy. Nevertheless, I think the solver is doing the right thing given the user input.

**Comment by wenegrat** [comment]
In case it is useful, you can see the computational setup we used in Wenegrat and Thomas 2020 [here.](https://wenegrat.github.io/publications/WenegratThomas2020.pdf). Equations 1-5 set up the equations we solve as perturbations away from a fixed background buoyancy gradient.

**Comment by navidcy** [comment]
Let me ask something: in the normal-g version, aren't we allowed to have z-direction = periodic?

**Comment by navidcy** [comment]
I think having periodicity in the dimension of gravity is not a problem conceptually. But it might indeed lead to big velocities that they may require smaller dt to tame the CFL. That's not a problem of the code though...

Similarly to, e.g., if you set up a highly unstable stratification then the code will blow up unless you choose your dt appropriately (see, e.g., #1190).


Now, in an attempt to show that having `z` periodic is not an issue, I tried to simulate the rising bubble with periodic boundary conditions. I got this puzzling result. Did I do anything wrong?

```Julia
using Oceananigans, Oceananigans.Grids, Plots

grid = RegularCartesianGrid(size = (1, 64, 64),
                            x = (0, 1), y = (-4, 4), z = (-4, 4),
                            topology = (Periodic, Periodic, Periodic))

model = IncompressibleModel(grid = grid,
                            architecture = CPU(),
                            advection = Oceananigans.Advection.WENO5(),
                            tracers = :b,
                            buoyancy = BuoyancyTracer(),
                            closure = IsotropicDiffusivity(ν=1e-3, κ=1e-3))

@info "Simulating a rising buoyant bubble with" model

set!(model, b = (x, y, z) -> exp(-y^2 - z^2))

simulation = Simulation(model, Δt=0.01, stop_iteration=1000)

using Oceananigans.OutputWriters

simulation.output_writers[:fields] = JLD2OutputWriter(model, (model.tracers.b,),
                                                      schedule = TimeInterval(0.1),
                                                      prefix = "bubble",
                                                      force = true)

run!(simulation)


using JLD2

file = jldopen(simulation.output_writers[:fields].filepath)

iterations = parse.(Int, keys(file["timeseries/t"]))

b = model.tracers.b

y, z = ynodes(b), znodes(b)


# Analyze the data

anim = @animate for (i, iteration) in enumerate(iterations)
  b_snapshot = file["timeseries/1/$iteration"][1, :, :]

  plt = contourf(y, z, b_snapshot',
                 xlabel = "y", ylabel = "z", title = "Buoyancy",
                 xlim = (grid.yF[1], grid.yF[end]), ylim = (grid.zF[1], grid.zF[end]),
                 aspectratio = :equal, linewidth = 0)
end

gif(anim, "bubble.gif", fps = 8) # hide
```

bubble

File: src/Buoyancy/gravitational_acceleration.jl
Comment (ali-ramadhan) [reply src/Buoyancy/gravitational_acceleration.jl:7]: Yeah I initially attempted to make `gravitational_acceleration` into a permanent 3-tuple but figured that using `g_x`, `g_y`, `g_z` would allow us to easily dispatch on `::Function` and `::Tuple{Function,Function,Function}` in the future in case we wanted g to vary with x, y, z.

**Comment by ali-ramadhan** [comment]
Thanks for the discussion @glwagner @tomchor @wenegrat I see that it was the initial condition that was unsuitable. Will read Wenegrat & Thomas (2020) to better understand how you set up your simulation there.

> Does this keep a resting fluid stratified in the direction of gravity at rest? John

Ah that would indeed be a good test: set up an inviscid ocean stratified in the direction of g and ensure that the velocities remain (close to) zero.

Ah interesting result @navidcy... Your setup looks good to me but not sure I can explain it either. Almost weirdly looks like the bubble is hitting an invisible boundary haha (I expected the bubble to keep rising and not spread out just yet...)

**Comment by glwagner** [comment]
> Did I do anything wrong?

@navidcy you may want to visualize the vertical velocity field.

It looks possibly like the right solution in the case that you had used a buoyancy field with zero domain mean (which it does not look like you did).  @ali-ramadhan can test this by tweaking his "sideways gravity" example to have zero initial mean buoyancy. @navidcy you can also test this by subtracting the domain mean from your initial buoyancy condition and confirming that you obtain the same result. You can also confirm that, no matter what mean buoyancy you set, you get the same result.

Finally, @navidcy I would check to see whether the hydrostatic pressure is vertically periodic in all of the cases you try.

**Comment by tomchor** [comment]
Am I missing something or does this currently just work with `SeawaterBuoyancy`? Are there plans to make this more general? (For example using `BuoyancyTracer`?)

**Comment by glwagner** [comment]
> Am I missing something or does this currently just work with `SeawaterBuoyancy`? Are there plans to make this more general? (For example using `BuoyancyTracer`?)

Looks like it to me. Within the constraints of the current design, I suppose that'd require adding a property to `BuoyancyTracer`. But you might not want to call it `gravitational_acceleration`, because that could be confusing. Something like `gravitational_direction` could work, with a default `(0, 0, 1)`. You'd also have to decide whether to normalize it or not in the constructor for `BuoyancyTracer`.

**Comment by tomchor** [comment]
Sorry if I'm missing something obvious (still new with the working of Oceananigans/Julia), but I was trying to test this feature and expand it to `BuoyancyTracer` over the weekend and I wonder if this implementation can be made more general. 

The way I understand @ali-ramadhan created some functions `g_x` to extract the gravity components in each direction and then used them in the `x_dot_g_b` functions to calculate each component of the gravity acceleration. The thing is that gravity can only be a vector using `SeawaterBuoyancy`. When trying to apply this to `BuoyancyTracer` I found that I'd need to replicate the same feature but with a different name (since, as @glwagner pointed out, `gravitational_acceleration` would be a confusing name when working directly with buoyancy).

Also, if you also think issue https://github.com/CliMA/Oceananigans.jl/issues/1372 is relevant, it might be nice to have one unified word/approach for the tilting of the domain.

So what do you guys think if this tilting is implement at the `AbstractBuoyancy` level? If I understand correctly all buoyancy types inherit that, right? That way we'd have a keyword for the tilt (maybe `tilt`, `gravitational_direction` or `gravity_projection`) that would default to `(0, 0, 1)` and a similar option could be given to `FPlane`, avoiding confusion.

**Comment by glwagner** [comment]
> So what do you guys think if this tilting is implement at the AbstractBuoyancy level? If I understand correctly all buoyancy types inherit that, right? That way we'd have a keyword for the tilt (maybe tilt, gravitational_direction or gravity_projection) that would default to (0, 0, 1) and a similar option could be given to FPlane, avoiding confusion.

Ah, the issue is that abstract types cannot have properties --- they can only be used to organize type parameters and dispatch.
The logic and motivation of your suggestion is sound though (avoiding code duplication through good design); we just have to come up with a different solution.

A similar solution could perhaps design a more hierarchical interface to `IncompressibleModel.buoyancy`. For example, we might write

```julia
struct Buoyancy{G, B}
    gravitational_direction :: G
    model :: B
end
```

The existing subtypes of `AbstractBuoyancy` are used for `Buoyancy.model`. The user API could be

```julia
buoyancy = Buoyancy(gravitational_direction=(0.1, 0, 0.9), model=BuoyancyTracer())
```

or

```julia
buoyancy = Buoyancy(gravitational_direction=(0.1, 0, 0.9),
                    model=SeawaterBuoyancy(gravitational_acceleration=9.81, equation_of_state=LinearEquationOfState(α=2e-4, β=8e-5)))
```

This is more verbose but could avoid some of the issues that @tomchor sees.

As for `Plane` and `BetaPlane`, I agree that those models are really predicated on a thin aspect ratio assumption that has to do with gravitational accelerations, and therefore "know" about the direction of gravity. We could either add a property `gravitational_direction` or, perhaps, add some wrapper / helper functions for constructing coriolis forces in tilted domains (eg `coriolis = TiltedCoriolisForces(FPlane(f=1e-4), vertical_direction=(0.1, 0.0, 0.9))` which returns `GeneralCoriolis` with 3D Coriolis forces --- or something).

**Comment by tomchor** [comment]
> buoyancy = Buoyancy(gravitational_direction=(0.1, 0, 0.9), model=BuoyancyTracer())

I actually really like this approach. It's a bit verbose, I agree, but it's extremely clear and intuitive. It seems like it would be a bit of work though, no? (I wouldn't know how to do it...)

**Comment by ali-ramadhan** [comment]
`gravitational_direction` for `Buoyancy` seems like a good name!

Is it a "direction" where g_x + g_y + g_z = 1 or is it a unit vector where |g| = g_x^2 + g_y^2 + g_z^2 = 1?

This PR has gotten a bit stale but should be easy to revive and add `gravitational_direction`.

I guess we thought this PR was missing a test? I'm happy to merge it as long as the 45 degree slanted rising thermal bubble looks good and we can worry about adding a more rigorous test in a future PR?

**Comment by francispoulin** [comment]
This all sounds very good and I don't have much to sugget but I wonder what people would think about `gravity_vector` instead of `gravitational_direction`?  My understanding is we are just specifying the vector that corresponds to gravity, but I might have missed something.

**Comment by tomchor** [comment]
I think it's more intuitive if `gravitational_direction` is a unit vector where `|g|=1`, no?

> I guess we thought this PR was missing a test? I'm happy to merge it as long as the 45 degree slanted rising thermal bubble looks good and we can worry about adding a more rigorous test in a future PR?

I ran the bubble case and it looked okay to me. I don't know if that's already good enough for merging this PR. I'd recomment merging once we get the more general tilt for buoyancy (if you agree that that's important.)

My idea was to start a more rigorous test using a tilted BBL example from the literature this weekend. I was thinking this could eventually become a docs example to show

- Tilted gravity
- Monin-Obukhov similarity wall model
- Maybe stretched grid if it's done by then?
- NetCDF outputs (I think none of the current examples use NetCDF)

I ended up not doing it because the papers I went through used both buoyancy as a tracer (not implement yet in this PR; hence my comment here) and tilt the domain in the `x` direction, so I'd need an `x` component of coriolis (hence my issue https://github.com/CliMA/Oceananigans.jl/issues/1372). I know these things are easy to circumvent with the available tools (i.e. using temperature to get the proper buoyancy and rotating the domain so that the tilt is in `y`, not `x`), but I thought it was best to make a thorough test once as these tools were in place already.

**Comment by francispoulin** [comment]
@tomchor , if I understand correclty, `'gravitational_direction` is to be a unit vector and `gravitational_acceleration` is the magnitude of gravity so the total acceleration due to gravity, say `\vec g` would be something like

``` 
\vec g = gravitational_acceleration * gravitational_direction

where |gravitational_direction| = 1?
```

What you are suggesting sounds very exciting and I for one am very interested to see it in action.

**Comment by ali-ramadhan** [comment]
Ah we should probably add @johncmarshall54's suggested test before merging: ensure an ocean stratified along the direction of g remains at rest.

**Comment by glwagner** [comment]
Certainly support `gravitational_direction` as a unit vector --- sorry if my non-unit example caused confusion.

**Comment by ali-ramadhan** [comment]
@tomchor and I added a `BuoyancyModel` type (can't call it `Buoyancy` since that would conflict with the module name, name is up for debate) to store the gravity unit vector and the buoyancy model.

Seems to work with `BuoyancyTracer` from @tomchor's simulations!

I guess what we weren't sure about is the user interface. Right now tilting gravity is done via e.g.

```julia
θ = 45
g = (0, sind(θ), cosd(θ))
model = IncompressibleModel(...
    buoyancy = BuoyancyModel(BuoyancyTracer(), gravity_unit_vector=g)
)
```

so now `model.buoyancy` is a `BuoyancyModel`.

---

But what should the user interface look like for not tilting gravity?

We could enforce explicit construction of a `BuoyancyModel` in every script, which may be seen as adding boilerplate, e.g.

```julia
model = IncompressibleModel(...
    buoyancy = BuoyancyModel(SeawaterBuoyancy())
)
```

or we could continue allowing

```julia
model = IncompressibleModel(...
    buoyancy = SeawaterBuoyancy()
)
```

and the model constructor can convert `SeawaterBuoyancy` to a `BuoyancyModel`. This is less boilerplate but maybe somewhat implicit.

I'm leaning towards the more implicit second solution of allowing just `buoyancy = SeawaterBuoyancy()`, but might be good to decide on one.

**Review by glwagner — COMMENTED** [review]
An alternative design would implement the functions `x_dot_g_b`, etc with

```julia
@inline x_dot_g_b(i, j, k, grid, buoyancy::BuoyancyModel, C) = @inbounds ĝ_x(buoyancy) * buoyancy_perturbation(i, j, k, grid, buoyancy.model, C)
```

while leaving the existing functionality of `buoyancy_perturbation` (which does appropriate dispatch on `buoyancy.model`) unchanged. The idea is that `buoyancy_perturbation` returns the buoyancy perturbation associated with `buoyancy_model`, while `ĝ_x` handles extracting the gravity unit vector.

This might require a less additional code and less dispatch foo than the current implementation.

I also suggest changing the module name to `Buoyancies` or `BuoyancyModels` and reserving `Buoyancy` for the user-facing object.

I also support something like

```julia
regularize_buoyancy(b::AbstractBuoyancy) = Buoyancy(b, gravity_unit_vector=(0, 0, 1))
```

or whatever's needed support the existing syntax of the `buoyancy` kwarg in the constructor for `IncompressibleModel`.

Does it also make sense to introduce a special type for `gravity_unit_vector=VerticalDirection()` with

```julia
x_dot_g_b(i, j, k, grid::AbstractGrid{FT}, ::Buoyancy{<:VerticalDirection}, C) where FT = zero(FT)
```

This will elide potentially expensive evaluations of `buoyancy_perturbation when they aren't needed (for example, the TEOS-10 buoyancy model involves 55 terms or something...)

**Comment by tomchor** [comment]
> I also suggest changing the module name to Buoyancies or BuoyancyModels and reserving Buoyancy for the user-facing object.

The current names that @ali-ramadhan and I used were always meant to be a placeholder as we couldn't agree on good ones to use. I quite like this suggestion (and I suspect @ali-ramadhan does too) so I implemented it.
Just as a reference, here are the commands that I used (plus the same for the struct `Buoyancy`):

```bash
find . -type f -name "*.jl" -exec sed -i "s/\<Buoyancy\>/BuoyancyModels/g" {} \;
find . -type f -name "*.md" -exec sed -i "s/\<Buoyancy\>/BuoyancyModels/g" {} \;
```

I also added an example in the sandbox that runs a fully quiescent fluid with a tilted stratification for an hour and then makes sure that the vertical and horizontal derivatives stay the way we expect them. (This test is done both for `BuoyancyTracer` and for `SeawaterBuoyancy`).

**Comment by ali-ramadhan** [comment]
@tomchor Thanks for adding the tilted gravity stratified fluid tests! I put them in `test_dynamics.jl` and they seem to pass even with 90 and 180 degree tilts!

I noticed that the test starts failing after ~10 hours for some angles though (or at least the deviations start becoming larger than square root of machine epsilon). For example:

```
N² * g̃[2] = 8.660254037844387e-6, mean(∂y_b) = 8.660254037844065e-6, Δ =  3.2187251995663413e-19 at t = 1 hour with θ=60
N² * g̃[3] = 5.0e-6,               mean(∂z_b) = 5.00000000000176e-6,  Δ = -1.7592874314471818e-18 at t = 1 hour with θ=60
N² * g̃[2] = 8.660254037844387e-6, mean(∂y_b) = 8.660130013623816e-6, Δ =  1.2402422057088938e-10 at t = 10 hours with θ=60
N² * g̃[3] = 5.0e-6,               mean(∂z_b) = 5.000172734522595e-6, Δ = -1.7273452259476338e-10 at t = 10 hours with θ=60

```

I guess the difference is still pretty small and this is a coarse 2D testing setup with low-order numerics, but it does grow quite a bit. Is this something to be worried about for turbulent simulations? I wonder if it's worth making a movie to see what happens after 10-100+ hours.

I think all tests should pass now so we could probably start using tilted gravity more seriously! Still need to address some of @glwagner's comments but will revisit this PR another day to do this.

cc @gianlisi @arnscheidt

**Comment by ali-ramadhan** [comment]
@tomchor suggested that the increased error might be due to errors from low-order numerics piling up over many iterations. Indeed, switching to WENO-5 + RK3 reduces the error by an order of magnitude!

```
N² * g̃[2] = 8.660254037844387e-6, mean(∂y_b) = 8.660130013623816e-6, Δ =  1.2402422057088938e-10 at t = 10 hours with θ=60 # CenteredSecondOrder + QAB2
N² * g̃[3] = 5.0e-6,               mean(∂z_b) = 5.000172734522595e-6, Δ = -1.7273452259476338e-10 at t = 10 hours with θ=60 # CenteredSecondOrder + QAB2
N² * g̃[2] = 8.660254037844387e-6, mean(∂y_b) = 8.660230615917627e-6, Δ =  2.3421926760073770e-11 at t = 10 hours with θ=60 # WENO5 + RK3
N² * g̃[3] = 5.0e-6,               mean(∂z_b) = 5.00003008535181e-6,  Δ = -3.0085351809812464e-11 at t = 10 hours with θ=60 # WENO5 + RK3
```

File: src/Grids/vertically_stretched_rectilinear_grid.jl
Comment (glwagner) [inline src/Grids/vertically_stretched_rectilinear_grid.jl:217]: 👀

Comment (tomchor) [reply src/Grids/vertically_stretched_rectilinear_grid.jl:217]: That was [your suggestion](https://github.com/CliMA/Oceananigans.jl/issues/1466#issuecomment-799796938)!

Comment (tomchor) [reply src/Grids/vertically_stretched_rectilinear_grid.jl:217]: Okay, just got warned that I committed to the wrong branch by mistake! Sorry

Comment (glwagner) [reply src/Grids/vertically_stretched_rectilinear_grid.jl:217]: Oh it's no great crime, provided you're ok with 👀 .  Everyone is tempted to scope creep every once in a while... :-P

**Comment by tomchor** [comment]
Does `regularize_buoyancy()` prevent old scripts using the old notation from failing?

**Comment by ali-ramadhan** [comment]
> Does `regularize_buoyancy()` prevent old scripts using the old notation from failing?

Yes, also helps reduce boilerplate due to e.g.

```julia
model = IncompressibleModel(...,
    buoyancy = Buoyancy(model=BuoyancyTracer())
)
```

since you can just do

```julia
model = IncompressibleModel(...,
    buoyancy = BuoyancyTracer()
)
```

But yeah also nice that there are no breaking changes.

**Comment by ali-ramadhan** [comment]
I think this PR is finally ready to be merged, provided that tests pass.

To make sure that `VerticalDirection` is working, I ran the equation of state benchmarks and they seem to match the benchmarks performed in https://github.com/CliMA/Oceananigans.jl/pull/1169#issuecomment-725471594

If anything, benchmarks seem a bit better (probably different machine) and `RoquetEquationOfState` is surprisingly faster on the CPU (but might be a fluke).

```
                                          Equation of state benchmarks
┌───────────────┬───────────────────────┬───────────┬───────────┬───────────┬───────────┬─────────────┬────────┐
│ Architectures │      EquationsOfState │       min │    median │      mean │       max │      memory │ allocs │
├───────────────┼───────────────────────┼───────────┼───────────┼───────────┼───────────┼─────────────┼────────┤
│           CPU │ LinearEquationOfState │   2.037 s │   2.040 s │   2.039 s │   2.041 s │  372.66 KiB │   2090 │
│           CPU │ RoquetEquationOfState │   1.759 s │   1.761 s │   1.761 s │   1.763 s │  373.77 KiB │   2090 │
│           CPU │ TEOS10EquationOfState │   2.270 s │   2.401 s │   2.378 s │   2.464 s │  372.53 KiB │   2090 │
│           GPU │ LinearEquationOfState │ 10.058 ms │ 13.161 ms │ 12.856 ms │ 13.215 ms │ 1022.19 KiB │   7154 │
│           GPU │ RoquetEquationOfState │ 10.688 ms │ 13.236 ms │ 12.991 ms │ 13.322 ms │    1.00 MiB │   7054 │
│           GPU │ TEOS10EquationOfState │ 10.204 ms │ 13.463 ms │ 13.145 ms │ 13.504 ms │ 1017.58 KiB │   7154 │
└───────────────┴───────────────────────┴───────────┴───────────┴───────────┴───────────┴─────────────┴────────┘
```

```
         Equation of state CPU -> GPU speedup
┌───────────────────────┬─────────┬─────────┬─────────┐
│      EquationsOfState │ speedup │  memory │  allocs │
├───────────────────────┼─────────┼─────────┼─────────┤
│ LinearEquationOfState │ 154.965 │ 2.74298 │ 3.42297 │
│ RoquetEquationOfState │ 133.062 │ 2.74052 │ 3.37512 │
│ TEOS10EquationOfState │ 178.317 │ 2.73152 │ 3.42297 │
└───────────────────────┴─────────┴─────────┴─────────┘
```

```
              Equation of state relative performance (CPU)
┌───────────────┬───────────────────────┬──────────┬──────────┬────────┐
│ Architectures │      EquationsOfState │ slowdown │   memory │ allocs │
├───────────────┼───────────────────────┼──────────┼──────────┼────────┤
│           CPU │ LinearEquationOfState │      1.0 │      1.0 │    1.0 │
│           CPU │ RoquetEquationOfState │ 0.863549 │  1.00298 │    1.0 │
│           CPU │ TEOS10EquationOfState │  1.17707 │ 0.999665 │    1.0 │
└───────────────┴───────────────────────┴──────────┴──────────┴────────┘
```

```
               Equation of state relative performance (GPU)
┌───────────────┬───────────────────────┬──────────┬──────────┬──────────┐
│ Architectures │      EquationsOfState │ slowdown │   memory │   allocs │
├───────────────┼───────────────────────┼──────────┼──────────┼──────────┤
│           GPU │ LinearEquationOfState │      1.0 │      1.0 │      1.0 │
│           GPU │ RoquetEquationOfState │  1.00569 │  1.00208 │ 0.986022 │
│           GPU │ TEOS10EquationOfState │  1.02292 │ 0.995491 │      1.0 │
└───────────────┴───────────────────────┴──────────┴──────────┴──────────┘

```

**Comment by tomchor** [comment]
> model = IncompressibleModel(...,
    buoyancy = Buoyancy(model=BuoyancyTracer())
)

I actually like the new, more verbose way better. It's way more explicit :)
But yeah, it's always good not to break stuff :+1:

**Review by tomchor — APPROVED** [review]

**Review by tomchor — APPROVED** [review]

**Comment by ali-ramadhan** [comment]
@glwagner I'm going to merge this since you're on vacation right now and seemed to generally agree with the direction of this PR on Zoom yesterday. I'm happy to address any leftover issues in a future PR.

**Comment by francispoulin** [comment]
I realize this PR has long since merged but was't sure where else to post.

I just saw a recent article that suggests that the horizontal compoments of gravity on Earth, which are 5 or 6 orders of magnitude smaller than the vertical, are important on the geostrophic motions.   Thanks to @ali-ramadhan and @tomchor who did a lot of work on this PR, this is a problem we could study easily in Oceananigans.

**Pull request merged**
